### DISCUSSION TITLE
Unsubscribe with onAuthStateChange

### DISCUSSION THREAD

**Discussion by jdgamble555** [supabase/supabase#5282] [Questions]
Is the the unsubscribe handled automatically with `onAuthStateChange`?  I don't see anything about observable memory leaks in the docs:

https://supabase.com/docs/reference/javascript/auth-onauthstatechange

How does this work, and should we be handling this?

J

**Comment by cureau** [comment]
Yea it looks like `onAuthStateChange` return `data` with an unsubscribe method (e.g., `ret.data.unsubscribe()`) ... might just need to be added to the docs.

Hopefully they don't change that setup in the future, cause we're going to use it in production. Subscribing to this thread just in case.

**Reply by bennik88** [reply]
Does it mean i should manually return something or not? I tried returning `res` but got an error `Warning: useEffect must not return anything besides a function, which is used for clean-up. You returned: [object Object]`

```js
useEffect(() => {
    const res = supabase.auth.onAuthStateChange((event, session) => {
    });
 
    // should I return here? return res.data.unsubscribe()
 }, []);
```

It seems like the `res` object has changes as my has the shape of:
Which would be `res.subscription.unsubscribe()`?
```js
{
    "subscription": {
        "id": "63a3b614-9446...",
        "callback": f (event, session),
         "unsubscribe": () => {...}
    },
    "error": null
}
```

It also seems to call `onAuthStateChange` more than 5 times on 'SIGNIN' and 'SIGNOUT'

**Reply by hjaber** [reply]
You should manually unsubscribe to prevent memory leaks. 

```
    const { data: authListener } = supabase.auth.onAuthStateChange(
      (event, session) => {
         console.log(event, session)
        }


    // Unsubscribe on cleanup
    return () => {
      authListener.unsubscribe();
    };
  }, []);
```

**Reply by jdgamble555** [reply]
Or, if you want to go crazy simple, I believe this should work as well:

```typescript
useEffect(() => supabase.auth.onAuthStateChange((event, session) => {
    //...
}).data?.unsubscribe, []);
```

J

**Reply by bennik88** [reply]
Thank you so much, guys. This community is key! 🚀 Just to update the answer, I'm using `v2 RC` of the Java Script Libary and for me it was.

```js
useEffect(() => {
    // Listen for auth changes on:
    // SIGNED_IN, SIGNED_OUT, TOKEN_REFRESHED, USER_UPDATED, USER_DELETED, PASSWORD_RECOVERY
    const { subscription } = supabase.auth.onAuthStateChange(
        async (event, session) => {
            console.log(event, session)
        }
    );
    // Unsubscribe on cleanup
    return () => {
      subscription?.unsubscribe();
    };
  }, []);
```

**Reply by hjaber** [reply]
> Thank you so much, guys. This community is key! 🚀 Just to update the answer, I'm using `v2 RC` of the Java Script Libary and for me it was.
> 
> ```js
> useEffect(() => {
>     // Listen for auth changes on:
>     // SIGNED_IN, SIGNED_OUT, TOKEN_REFRESHED, USER_UPDATED, USER_DELETED, PASSWORD_RECOVERY
>     const { subscription } = supabase.auth.onAuthStateChange(
>         async (event, session) => {
>             console.log(event, session)
>         }
>     );
>     // Unsubscribe on cleanup
>     return () => {
>       subscription?.unsubscribe();
>     };
>   }, []);
> ```

on supabase/supabase-js V2 RC, `subscription` does not exist as a property for `onAuthStateChange`

It is currently `const = { data: { subscription } } = supabase.auth.onAuthStateChange(...)`

**Reply by bennik88** [reply]
That's quite interesting my return object is as stated above:
With ` "@supabase/supabase-js": "^2.0.0-rc.2",` from my package.json
```js
{
    "subscription": {
        "id": "63a3b614-9446...",
        "callback": f (event, session),
         "unsubscribe": () => {...}
    },
    "error": null
}
```

**Comment by hjaber** [comment]
You are correct, on `@supabase/supabase-js": "^2.0.0-rc.2`, that does work.

However, a change was made on Aug 20 2022 that will no longer support that:
[
GoTrueClient.ts commit history ](https://github.com/supabase/gotrue-js/commit/07c3cb2aaa610d2c854955fbd4615f8edf14d99d)

If you upgrade to the latest version of supabase/supabase-js V2 RC, you will need to make the changes above.

**Reply by bennik88** [reply]
Thanks for the news. Just updated the package and now the change to ` const { data: { subscription } } = supabase.auth.onAuthStateChange(...)` is visible as described.
